croissant/online_training_helper.py

```
@@ -0,0 +1,94 @@
+import json
+import logging
+import mlflow
+import multiprocessing
+
+
+class EnvHelperException(Exception):
+    pass
+
+
+class EnvHelperSchema(argschema.ArgSchema):
+    stack = argschema.fields.Str(
+        required=True,
+        description="stack name from which to suggest environment variables")
+    timeout = argschema.fields.Float(
+        required=False,
+        default=20.0,
+        description=("timeout for mlflow tracking client list_experiments "
+                     "call. Without a VPN connection, this call hangs."))
+
+
+def mlflow_experiment_check(tracking_uri):
+    client = mlflow.tracking.MlflowClient(tracking_uri=tracking_uri)
+    experiments = client.list_experiments()
+    exoptions = ""
+    for e in experiments:
+        if e.artifact_location.startswith("s3://"):
+            exoptions += (f"\nexport MLFLOW_EXPERIMENT={e.name}")
+    return exoptions
+
+
+class EnvHelper(argschema.ArgSchemaParser):
+    default_schema = EnvHelperSchema
+
+    def run(self):
+        client = boto3.client('cloudformation')
+        try:
+            desc = client.describe_stacks(StackName=self.args['stack'])
+        except NoCredentialsError as e:
+            logging.error(str(e))
+            logging.error("no AWS credentials found.")
+            return
+
+        if len(desc['Stacks']) != 1:
+            raise EnvHelperException(f"description for {self.args['stack']} "
+                                     "did not return exactly one entry")
+        outs = {i['OutputKey']: i['OutputValue']
+                for i in desc['Stacks'][0]['Outputs']}
+
+        exports = ""
+        lookup = {
+                'ONLINE_TRAINING_CLUSTER': outs['FargateClusterName'],
+                'ONLINE_TRAINING_TASK': outs['TaskDefinitionName'],
+                'ONLINE_TRAINING_CONTAINER': outs['ContainerNameOutput'],
+                'ONLINE_TRAINING_SUBNET': outs['PrivateSubnet'],
+                'ONLINE_TRAINING_SECURITY': outs['VPCUserSecurityGroupId']}
```

Can you explain VpcUserSecurityGroup over VpcSecurityGroup here?

---

I thought of it as: the container will want to access resources, users can access resources, let's put it in that group. I tried the other way and launched jobs failed silently. I'm inclined to leave it this way, unless it is a security issue.

---

LGTM as discussed offline. We should test with another user's IP to make sure it's ok.